File: frontend/src/live/transcript.ts
import type { TranscriptEntry } from "./state";


export const CONVERSATION_TRANSCRIPT_TOPIC = "voice-conversation.transcript.v1";

export interface ConversationTranscriptUpdate {
  version: 1;
  attemptId: string;
  sequence: number;
  emittedAt: string;
  entry: TranscriptEntry;
}

export function parseConversationTranscriptUpdate(
  payload: Uint8Array,
): ConversationTranscriptUpdate | null {
  try {
    const parsed: unknown = JSON.parse(new TextDecoder().decode(payload));
    if (!isRecord(parsed) || parsed.version !== 1) return null;
    if (typeof parsed.attemptId !== "string" || !parsed.attemptId.trim()) return null;
    if (!Number.isInteger(parsed.sequence) || Number(parsed.sequence) < 1) return null;
    if (
      typeof parsed.emittedAt !== "string" ||
      !parsed.emittedAt ||
      Number.isNaN(Date.parse(parsed.emittedAt))
    ) return null;
    if (!isTranscriptEntry(parsed.entry)) return null;
    return {
      version: 1,
      attemptId: parsed.attemptId,
      sequence: Number(parsed.sequence),
      emittedAt: parsed.emittedAt,
      entry: {
        id: parsed.entry.id,
        role: parsed.entry.role,
        text: parsed.entry.text.trim(),
        final: parsed.entry.final,
      },
    };
  } catch {
    return null;
  }
}

function isTranscriptEntry(value: unknown): value is TranscriptEntry {
  return (
    isRecord(value) &&
    typeof value.id === "string" &&
    Boolean(value.id.trim()) &&
    (value.role === "user" || value.role === "agent") &&
    typeof value.text === "string" &&
    Boolean(value.text.trim()) &&
    typeof value.final === "boolean"
  );
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

File: backend/src/voice_presentation/transport/transcript.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator
from pydantic.alias_generators import to_camel


CONVERSATION_TRANSCRIPT_TOPIC = "voice-conversation.transcript.v1"


class ConversationTranscriptEntry(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str = Field(min_length=1)
    role: Literal["user", "agent"]
    text: str = Field(min_length=1)
    final: bool

    @field_validator("id", "text")
    @classmethod
    def strip_non_blank_text(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("transcript fields cannot be blank")
        return value


class ConversationTranscriptUpdate(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="forbid",
        frozen=True,
    )

    version: Literal[1] = 1
    attempt_id: str = Field(min_length=1)
    sequence: int = Field(ge=1)
    emitted_at: datetime
    entry: ConversationTranscriptEntry

    @field_validator("attempt_id")
    @classmethod
    def strip_non_blank_attempt_id(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("attempt_id cannot be blank")
        return value

    @classmethod
    def from_entry(
        cls,
        *,
        attempt_id: str,
        sequence: int,
        entry: ConversationTranscriptEntry,
    ) -> "ConversationTranscriptUpdate":
        return cls(
            attempt_id=attempt_id,
            sequence=sequence,
            emitted_at=datetime.now(UTC),
            entry=entry,
        )

    def to_json(self) -> str:
        return self.model_dump_json(by_alias=True)
